### PR TITLE
Integration test for cook scheduler / rebalancer using mock mesos

### PR DESCRIPTION
**Pull request by wyegelwel**
This PR includes a mock of mesos and a simulator that starts the scheduling aspects of cook to allow back testing and auditing behavior. The mock of mesos allows us to simulate arbitrarily large clusters (until computation or memory limits on the machine running the simulation are reached). The intent is this simulator will be run "faster than real time" so that running over a day or week of historical jobs can be done in minutes and hours.

### CODE REVIEW

File: scheduler/src/cook/components.clj [hunk 1]
```diff
@@ -84,35 +84,38 @@
...
-                            rebalancer
-                            {:fenzo-max-jobs-considered fenzo-max-jobs-considered
-                             :fenzo-scaleback fenzo-scaleback
-                             :fenzo-floor-iterations-before-warn fenzo-floor-iterations-before-warn
-                             :fenzo-floor-iterations-before-reset fenzo-floor-iterations-before-reset
-                             :good-enough-fitness good-enough-fitness})
-                           (catch ClassNotFoundException e
-                             (log/warn e "Not loading mesos support...")
-                             nil)))})
+                         (let [make-mesos-driver-fn (partial (lazy-load-var 'cook.mesos/make-mesos-driver)
```
Comment (dposada) [inline scheduler/src/cook/components.clj]: Why is the `lazy-load-var` necessary here?

File: scheduler/src/cook/mesos/mesos_mock.clj [hunk 2]
```diff
@@ -0,0 +1,458 @@
...
+                    [state' new-offer-chan]))]
+            (when state'
+              (recur state' new-offer-chan))))
+        (catch Exception ex
+          (log/fatal ex "Error while simulating mesos")
+          (println "Exception occurred: " ex))))
+    (mesos/wrap-driver driver)))
+
+
+(comment
```
Comment (dposada) [inline scheduler/src/cook/mesos/mesos_mock.clj]: Can we remove the comment?

File: scheduler/test/cook/test/mesos/mesos_mock.clj [hunk 3]
```diff
@@ -0,0 +1,749 @@
+(ns cook.test.mesos.mesos-mock
+  (:use clojure.test)
+  (:require [cook.mesos :as c]
```
Comment (dposada) [inline scheduler/test/cook/test/mesos/mesos_mock.clj]: Sort requires

File: scheduler/test/cook/test/mesos/mesos_mock.clj [hunk 4]
```diff
@@ -0,0 +1,749 @@
...
+                       )
+    (disconnected [this driver]
+                  )
+    ;; We don't care about losing slaves or executors--only tasks
+    (slave-lost [this driver slave-id])
+    (executor-lost [this driver executor-id slave-id status])
+    (error [this driver message]
+           )
+    (resource-offers [this driver offers]
+                     )
```
Comment (dposada) [inline scheduler/test/cook/test/mesos/mesos_mock.clj]: Several `)` that can be on the previous line

**Comment by dposada** [comment]
@wyegelwel I see several `TODO`s in the code - are those things you want to do as part of this PR or just some time in the future?

File: scheduler/src/cook/components.clj (hunk 1 above)
Comment (wyegelwel) [inline scheduler/src/cook/components.clj]: The `lazy-load-var` are a way to avoid requiring namespaces initially which has two consequences (I explicitly didn't use "benefits" here):
1. AOT compilation can pass even if the lazy loaded namespace can't be compiled (since it is done at run time)
2. It makes initial start up time faster since not everything is loaded immediately

I'm not convinced either of these are worth the headache but I don't think this CR is the place to address it.

File: scheduler/src/cook/mesos.clj [hunk 5]
```diff
@@ -75,9 +75,73 @@
...
+            (when framework-id
+              {:id {:value framework-id}}))
+          mesos-master
+          (when mesos-principal
+            [{:principal mesos-principal}]))))
+
 (defn start-mesos-scheduler
-  "Starts a leader elector that runs a mesos."
-  [mesos-master mesos-master-hosts curator-framework mesos-datomic-conn mesos-datomic-mult zk-prefix mesos-failover-timeout mesos-principal mesos-role mesos-framework-name offer-incubate-time-ms mea-culpa-failure-limit task-constraints riemann-host riemann-port mesos-pending-jobs-atom gpu-enabled? rebalancer-config
+  "Starts a leader elector which while leader starts the mesos scheduler and
```
Comment (wyegelwel) [inline scheduler/src/cook/mesos.clj]: me either, will fix

File: scheduler/src/cook/mesos/mesos_mock.clj [hunk 6]
```diff
@@ -0,0 +1,458 @@
+(ns cook.mesos.mesos-mock
+  (:require [mesomatic.scheduler :as mesos]
```
Comment (wyegelwel) [inline scheduler/src/cook/mesos/mesos_mock.clj]: Just kidding, of course

**Comment by wyegelwel** [comment]
Ok, this PR includes code to mock mesos and have the cook scheduler connect to it. It includes a test of the basic schedule / rebalance workflow that was previously missing from our tests. 

There will be additional changes to make a full-fledge simulator but I think this is a milestone to merge. 

@dposada when you get a moment, could you take another pass? 

The todos that are remaining are for future work

**Comment by dposada** [comment]
Will do

File: scheduler/src/cook/mesos/mesos_mock.clj [hunk 7]
```diff
@@ -0,0 +1,454 @@
...
+        slave-id (:value (:slave-id (first tasks)))]
+    (when-not (every? #(= slave-id (:value (:slave-id %))) tasks)
+      (throw (ex-info "All tasks must have same slave-id" 
+                      {:slave-ids (map (comp :value :slave-id) tasks)})))
+    (when-not (get slave-id->host slave-id)
+      (throw (ex-info "Unknown slave-id" {:slave-id slave-id :known-slave-ids (keys slave-id->host)})))
+    (when (some nil? offers) 
+      (throw (ex-info "Unknown offers" 
+                      {:unknown-offers 
+                       (remove #(contains? (set (keys offer-id->offer)))
```
Comment (dposada) [inline scheduler/src/cook/mesos/mesos_mock.clj]: Should this be 

```clj
#(contains? (set (keys offer-id->offer)) %) 
```
?

File: scheduler/test/cook/test/mesos/mesos_mock.clj [hunk 8]
```diff
@@ -0,0 +1,844 @@
...
+              update-status-error-ms 200
+              ]
+          ;; Transact user "a"s jobs
+          (doall user-a-job-ent-ids)
+          ;; Let them get scheduled
+          (Thread/sleep 5000)
+          ;; Transact user "b"s jobs
+          (doall user-b-job-ent-ids)
+          ;; Let the system complete
+          (Thread/sleep 30000)
```
Comment (dposada) [inline scheduler/test/cook/test/mesos/mesos_mock.clj]: Is it possible to refactor this to avoid the `sleep`s and instead poll for expected outcomes with a timeout?

File: scheduler/test/cook/test/mesos/mesos_mock.clj [hunk 9]
```diff
@@ -0,0 +1,844 @@
...
+     (.start curator-framework)
+     [zookeeper-server curator-framework])))
+
+(defmacro with-cook-scheduler
+  [conn make-mesos-driver-fn
+   {:keys [get-mesos-utilization mesos-datomic-mult zk-prefix offer-incubate-time-ms
+           mea-culpa-failure-limit task-constraints riemann-host riemann-port
+           mesos-pending-jobs-atom gpu-enabled? rebalancer-config fenzo-config]
+    :or {get-mesos-utilization (fn [] 0.9) ; can get this from mesos mock later
+;         mesos-datomic-mult (async/mult (async/chan)) ; don't use mult for now
```
Comment (dposada) [inline scheduler/test/cook/test/mesos/mesos_mock.clj]: Can we delete this commented line?

File: scheduler/test/cook/test/mesos/mesos_mock.clj [hunk 10]
```diff
@@ -0,0 +1,844 @@
...
+                  instance (first instances)]
+              (is (= (count instances) 1))
+              (is (= (:job/state job-ent) :job.state/completed ))
+              (log/info "job i: " (d/touch job-ent))
+              (is (< 0
+                     (- (.getTime (:instance/end-time instance))
+                        (.getTime (:instance/start-time instance))
+                        (read-string (:job/command job-ent)))
+                     update-status-error-ms))))
+          (dump-jobs-to-csv (d/db mesos-datomic-conn) "trace.csv"))))))
```
Comment (dposada) [inline scheduler/test/cook/test/mesos/mesos_mock.clj]: Comments/questions on the csv creation:

- Should we `log/debug` the trace info as well?
- Can we make the csv file optional? I suspect we usually won't care about the csv file
- Can we name the file more descriptively e.g. `trace-basic-scheduling-rebalancing-test.csv`
- Can we add an ignore for `scheduler/trace*.csv`

File: scheduler/test/cook/test/mesos/mesos_mock.clj [hunk 11]
```diff
@@ -0,0 +1,844 @@
...
+            (let [job-ent (d/entity (d/db mesos-datomic-conn) job-ent-id)
+                  instances (:job/instance job-ent)
+                  instance (first instances)]
+              (is (= (count instances) 1))
+              (is (= (:job/state job-ent) :job.state/completed ))
+              (log/info "job i: " (d/touch job-ent))
+              (is (< 0
+                     (- (.getTime (:instance/end-time instance))
+                        (.getTime (:instance/start-time instance))
+                        (read-string (:job/command job-ent)))
```
Comment (dposada) [inline scheduler/test/cook/test/mesos/mesos_mock.clj]: Why are we calling `read-string` on the command?

File: scheduler/test/cook/test/mesos/mesos_mock.clj [hunk 12]
```diff
@@ -0,0 +1,844 @@
...
+                                     :mem {}
+                                     :ports {}}
+               :slave-id "a5cb3212-3023-4546-888d-67cc13eac3dd"}}))
+      (is (= (count (vals (:offer-id->offer new-state))) 2)))))
+
+(deftest combine-ranges
+  (testing "basic test"
+    (is (= (mm/combine-ranges [{:begin 16 :end 20} {:begin 0 :end 2}
+                               {:begin 3 :end 5} {:begin 8 :end 10}]))
+        [{:begin 0 :end 5} {:begin 8 :end 10} {:begin 16 :end 20}])))
```
Comment (wyegelwel) [inline scheduler/test/cook/test/mesos/mesos_mock.clj]: That is super embarassing, good catch

File: scheduler/test/cook/test/mesos/mesos_mock.clj [hunk 13]
```diff
@@ -0,0 +1,844 @@
...
+          scheduler (mesos/scheduler
+                      (registered [this driver framework-id master-info]
+                                  (reset! registered-atom true))
+                      (resource-offers [this driver offers]
+                                       (swap! offer-atom into offers )))
+          hosts [(make-random-host)]
+          speed-multiplier 20
+          mock-driver (mm/mesos-driver-mock hosts speed-multiplier scheduler)]
+      (mesos/start! mock-driver)
+      (Thread/sleep 300) ;; Wait for offers to be sent
```
Comment (wyegelwel) [inline scheduler/test/cook/test/mesos/mesos_mock.clj]: Done, thanks this was a great suggestion

Comment (wyegelwel) [inline scheduler/test/cook/test/mesos/mesos_mock.clj]: (and by done, I mean when I push next it will be there)

File: scheduler/test/cook/test/mesos/mesos_mock.clj (hunk 8 above)
Comment (wyegelwel) [inline scheduler/test/cook/test/mesos/mesos_mock.clj]: This one is a little trickier to remove because of the polling that would need to be done. I used the tracing csv to make close bounds so it shouldn't be too much slower. In tests, I was seeing things complete in ~22-25 seconds and so gave a little bit of a buffer. Can we push this off for a subsequent PR?

File: scheduler/test/cook/test/mesos/mesos_mock.clj (hunk 11 above)
Comment (wyegelwel) [inline scheduler/test/cook/test/mesos/mesos_mock.clj]: The int of the command is the run time. It is the default way to set the run time of a task with the mesos mock. I agree this is a little sketchy.

File: scheduler/test/cook/test/mesos/mesos_mock.clj (hunk 10 above)
Comment (wyegelwel) [inline scheduler/test/cook/test/mesos/mesos_mock.clj]: Yea, I added this because there was a bug with the integration and I couldn't figure it out without the trace. I will have it write only when log is set debug, rename, and add the ignore. 

Going to skip log/debug because I think it will look noisy in the log.

File: scheduler/src/cook/mesos.clj [hunk 14]
```diff
@@ -75,9 +78,90 @@
...
+              {:failover-timeout mesos-failover-timeout})
+            (when framework-id
+              {:id {:value framework-id}}))
+          mesos-master
+          (when mesos-principal
+            [{:principal mesos-principal}]))))
+
+(defn get-mesos-utilization
+  [mesos-master-hosts]
+  (let [mesos-master-urls (map #(str "http://" % ":5050/metrics/snapshot") mesos-master-hosts)
```
Comment (wyegelwel) [inline scheduler/src/cook/mesos.clj]: Yes but I copied this from cook.mesos.scheduler as is. Let's make it configurable in a different PR

File: scheduler/src/cook/mesos/mesos_mock.clj (hunk 7 above)
Comment (wyegelwel) [inline scheduler/src/cook/mesos/mesos_mock.clj]: Yes, thanks for catching that. I added these checks to save me grief when there was a bug upstream but I guess I haven't exercised that one yet.

File: scheduler/test/cook/test/mesos/mesos_mock.clj (hunk 8 above)
Comment (wyegelwel) [inline scheduler/test/cook/test/mesos/mesos_mock.clj]: Never mind, I did it

**Comment by wyegelwel** [comment]
Ok, just pushed updates. Thanks for the suggestions, seeing big improvements.

File: scheduler/src/cook/mesos/mesos_mock.clj [hunk 15]
```diff
@@ -0,0 +1,430 @@
...
+                                                    :version "1.0.1"}))
+         (loop [state {:config {:task->runtime-ms task->runtime-ms
+                                :task->complete-status task->complete-status
+                                :speed-multiplier speed-multiplier}
+                       :slave-id->host (map-from-vals :slave-id hosts)
+                       :offer-id->offer {}
+                       :task-id->task {}
+                       :task-id->completed-task {} ;;TODO test this in kill
+                       :complete-chan->task-id {}}
+                new-offer-chan (async/timeout (/ 1000 speed-multiplier))]
```
Comment (dposada) [inline scheduler/src/cook/mesos/mesos_mock.clj]: Since we have `(/ 1000 speed-multiplier)` duplicated here and below, can we move it up into the outer `let` and call it something like `new-offer-delay-ms`?

File: scheduler/src/cook/mesos/mesos_mock.clj [hunk 16]
```diff
@@ -0,0 +1,430 @@
...
+   `scheduler` is an implementation of the mesomatic scheduler protocol"
+  ([hosts speed-multiplier scheduler]
+   (mesos-driver-mock hosts speed-multiplier default-task->runtime-ms scheduler))
+  ([hosts speed-multiplier task->runtime-ms scheduler]
+   (mesos-driver-mock hosts speed-multiplier task->runtime-ms default-task->complete-status scheduler))
+  ([hosts speed-multiplier task->runtime-ms task->complete-status scheduler]
+   (let [action-chan (async/chan 10) ; This will block calls to the driver. This may be problematic..
+         exit-chan (async/chan)
+         start-chan (async/chan)
+         driver (reify SchedulerDriver
```
Comment (dposada) [inline scheduler/src/cook/mesos/mesos_mock.clj]: Perhaps move this `reify` out into a separate function

File: scheduler/test/cook/test/mesos/mesos_mock.clj [hunk 17]
```diff
@@ -0,0 +1,844 @@
...
+    (framework-message [this driver executor-id slave-id data])
+    (disconnected [this driver])
+    (slave-lost [this driver slave-id])
+    (executor-lost [this driver executor-id slave-id status])
+    (error [this driver message])
+    (resource-offers [this driver offers])
+    (status-update [this driver status])))
+
+(deftest make-offer
+  (let [basic-host-info {:hostname "hosta"
```
Comment (dposada) [inline scheduler/test/cook/test/mesos/mesos_mock.clj]: Should we create a helper function for creating these host info maps? We've got them in several places and they're kind of big.

File: scheduler/test/cook/test/mesos/mesos_mock.clj [hunk 18]
```diff
@@ -0,0 +1,844 @@
...
+            :attributes {}
+            :resources {:cpus {"*" 12}
+                        :mem {"*" 120000}
+                        :ports {"*" [{:begin 10000 :end 30000}]}}
+            :available-resources {:cpus {"*" 4.0}
+                                  :mem {"*" 40000.0}
+                                  :ports {"*" [{:begin 10000 :end 12000} {:begin 25001 :end 30000}]}}
+            :slave-id "a5cb3212-3023-4546-888d-67cc13eac3dd"}))))
+
+(defn make-random-host
```
Comment (dposada) [inline scheduler/test/cook/test/mesos/mesos_mock.clj]: `make-random-host` sort of implies that the contents of the host are random, when really it's only the `uuid` and `slave-id` that are. Maybe `dummy-host`?

File: scheduler/test/cook/test/mesos/mesos_mock.clj [hunk 19]
```diff
@@ -0,0 +1,844 @@
...
+          (poll-until #(= (count @task-complete-atom) 2) 20 1000)
+          (mesos/stop! mock-driver)
+          (is @registered-atom)
+          (is (= (count @launched-atom) 2))
+          (is (= (count @task-running-atom) 2))
+          (is (= (count @task-complete-atom) 2)))
+        (catch Throwable t
+          (throw (ex-info "Error while testing launch" atom-map t)))))))
+
+
```
Comment (dposada) [inline scheduler/test/cook/test/mesos/mesos_mock.clj]: Nitpick: extra newline

File: scheduler/test/cook/test/mesos/mesos_mock.clj [hunk 20]
```diff
@@ -0,0 +1,844 @@
...
+  ([zk-port]
+   ;; Copied from src/cook/components
+   ;; TODO: don't copy from components
+   (let [retry-policy (BoundedExponentialBackoffRetry. 100 120000 10)
+         ;; The 180s session and 30s connection timeouts were pulled from a google group
+         ;; recommendation
+         zookeeper-server (org.apache.curator.test.TestingServer. zk-port false)
+         zk-str (str "localhost:" zk-port)
+         curator-framework (CuratorFrameworkFactory/newClient zk-str 180000 30000 retry-policy)
+         ]
```
Comment (dposada) [inline scheduler/test/cook/test/mesos/mesos_mock.clj]: `]` can be on previous line

File: scheduler/test/cook/test/mesos/mesos_mock.clj [hunk 21]
```diff
@@ -0,0 +1,844 @@
...
+                                                     :user "b"
+                                                     :command "1000"
+                                                     :custom-executor? false
+                                                     :memory mem
+                                                     :ncpus cpus))
+              update-status-error-ms 500]
+          ;; Transact user "a"s jobs
+          (doall user-a-job-ent-ids)
+          ;; Let them get scheduled
+          (Thread/sleep 5000)
```
Comment (dposada) [inline scheduler/test/cook/test/mesos/mesos_mock.clj]: Can we use `poll-until` here as well?

File: scheduler/test/cook/test/mesos/mesos_mock.clj (hunk 17 above)
Comment (wyegelwel) [inline scheduler/test/cook/test/mesos/mesos_mock.clj]: I will change where it makes sense. For this function, the offer is built based on the values of the host so we would need to specify everything about the host anyway.

File: scheduler/src/cook/mesos.clj [hunk 22]
```diff
@@ -75,9 +78,92 @@
...
+              {:id {:value framework-id}}))
+          mesos-master
+          (when mesos-principal
+            [{:principal mesos-principal}]))))
+
+(defn get-mesos-utilization
+  "Queries the mesos master to get the utilization.
+   Returns the max of cpu and mem utilization as a decimal (e.g. 0.92)"
+  [mesos-master-hosts]
+  (let [mesos-master-urls (map #(str "http://" % ":5050/metrics/snapshot") mesos-master-hosts)
```
Comment (wyegelwel) [inline scheduler/src/cook/mesos.clj]: Sorry, thought I responded to this last time. Yes, but I don't think we should in this PR. This function was moved from another namespace (cook.mesos.rebalancer) because it was nicer from an injection standpoint to have it in cook.mesos. However, changing the functionality of it is unrelated to this pr.

File: scheduler/test/cook/test/mesos/mesos_mock.clj (hunk 17 above)
Comment (wyegelwel) [inline scheduler/test/cook/test/mesos/mesos_mock.clj]: I ended up not changing this anywhere. My reasoning is, for the tests that have the structure shown explicitly, the test will later make assertions based on the particular data in the structure. Specifically, the problem is all the functions are mutating the structure in some way so writing a generic "thing is what I expect function" would be effectively the same as the function I'm trying to test. 

The alternative would be to def the host map but in that case, a change to a value for one test will end up breaking every other test. Also, it would be less clear why the values are the way the are because they are defined elsewhere.

I'm not sure I care so much that if you feel strong about def-ing the host I will do it.

File: scheduler/src/cook/mesos/mesos_mock.clj [hunk 23]
```diff
@@ -0,0 +1,430 @@
...
+(defn mesos-driver-mock
+  "A mock mesos implementation.
+
+   It stores the 'hosts' in the cluster, how much resources are available on
+   each host, what jobs are running on each 'host' and how much longer each
+   job is expected to run for.
+
+   Parameters:
+   `hosts` is a seq of maps that contain a hostname, agent id, resources and attributes
+   `speed-multiplier` is a number >1 which is how much to speed up 'time'
```
Comment (wyegelwel) [inline scheduler/src/cook/mesos/mesos_mock.clj]: The speed multiplier isn't really used yet but the intent with all of this is to write a faster than real time simulator where we can play back what happened in a production cluster in order to test improvements to the scheduling algorithm. The fact that it helps us improve our testing is an added bonus.

In the simulator, being about to pick the speed up factor is important to compare different runs of the simulation.

File: scheduler/test/cook/test/mesos/mesos_mock.clj (hunk 11 above)
Comment (dposada) [inline scheduler/test/cook/test/mesos/mesos_mock.clj]: Response from @wyegelwel:

> Yea, probably but I'm not sure it is worth changing. For the simulator, I will set something that isn't the default and the simulator is really the only use case for this

File: scheduler/test/cook/test/mesos/mesos_mock.clj (hunk 17 above)
Comment (dposada) [inline scheduler/test/cook/test/mesos/mesos_mock.clj]: I completely agree with your points. 

I was thinking something more along the lines of:

```clj
(defn host-info
  "Constructs a host-info map with the given data"
  ([hostname slave-id cpus mem ports]
   (host-info hostname slave-id cpus mem ports nil nil nil))
  ([hostname slave-id cpus mem [ports-begin ports-end] cpus-available
    mem-available [ports-available-begin ports-available-end]]
   {:hostname hostname
    :attributes {}
    :resources {:cpus {"*" cpus}
                :mem {"*" mem}
                :ports {"*" [{:begin ports-begin :end ports-end}]}}
    :available-resources (cond-> {:cpus {}, :mem {}, :ports {}}
                                 cpus-available (assoc :cpus {"*" cpus-available})
                                 mem-available (assoc :mem {"*" mem-available})
                                 (and ports-available-begin
                                      ports-available-end) (assoc :ports {"*" [{:begin ports-available-begin
                                                                                :end ports-available-end}]}))
    :slave-id slave-id}))
```

which you could then use in tests, either with available resources or without:

```clj
;; with available resources
(host-info "hosta" "e8ab3892-3023-4546-888d-67cc13eac3cc" 10 100000 [10000 20000] 10 100000 [10000 20000])

;; without
(host-info "hostb" "a5cb3212-3023-4546-888d-67cc13eac3dd" 12 120000 [10000 20000])
```

This avoids the "thing that I expect" function and the host map `def`, while also preventing a bunch of boilerplate and making the tests proper easier to read. What do you think?

Comment (wyegelwel) [inline scheduler/test/cook/test/mesos/mesos_mock.clj]: This removes the boilerplate but then requires looking back at this function to remember what the args mean. I first tried switching to use the `dummy-host` function but it just changed one type of boilerplate for another. 

I propose we merge as is so that we can get the benefit of the tests and I can avoid the code drifting further. I will be continuing to work on this code to build out the simulator and will keep the idea of reducing the boilerplate in mind. Perhaps you can help me brainstorm how to do this out of band.
